src/ai/backend/manager/server.py

```diff
@@ -490,6 +491,113 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
         await ectx.cleanup()
 
 
+@actxmgr
+async def hanging_sessions_scanner_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    from contextlib import suppress
+    from datetime import timedelta
+    from typing import TYPE_CHECKING
+
+    import sqlalchemy as sa
+    from dateutil.tz import tzutc
+    from sqlalchemy.orm import load_only, noload
+
+    from .models.session import SessionStatus
+
+    if TYPE_CHECKING:
+        from .models.utils import ExtendedAsyncSAEngine
+
+    async def _fetch_hanging_sessions(
+        db: ExtendedAsyncSAEngine,
+        status: SessionStatus,
+        threshold: timedelta,
+    ) -> tuple[SessionRow, ...]:
+        query = (
+            sa.select(SessionRow)
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+        reason: KernelLifecycleEventReason = KernelLifecycleEventReason.HANGTIME_EXCEEDED,
+    ) -> None:
+        while True:
+            sessions = await _fetch_hanging_sessions(
+                root_ctx.db, status=status, threshold=threshold
+            )
+            log.debug(f"{len(sessions)} {status} sessions found.")
+
+            _ = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(session, forced=True, reason=reason),
+                    )
+                    for session in sessions
+                ]
+            )
+
+            await asyncio.sleep(threshold.seconds)
+
+    raw_session_config = await root_ctx.shared_config.etcd.get_prefix("config/session")
+    assert isinstance(raw_session_config, Mapping)
+
+    hang_tolerance_threshold_dict = raw_session_config.get("hang-tolerance-threshold", {})
+    assert isinstance(hang_tolerance_threshold_dict, Mapping)
```

Comment:
We can avoid this boilerplate by splitting out the inner `t.Dict()` trafaret in manager/config.py as a separate top-level variable to reuse it in both the whole `shared_config` input validation and a typed deserializer when reading the etcd on the fly.

Reply:
Resolved: c666327407e85fc27dc5dabd6cf2e426b318af3d